hls-plugin-api/src/Ide/PluginUtils.hs

```diff
@@ -263,27 +277,64 @@ allLspCmdIds pid commands = concatMap go commands
 
 -- ---------------------------------------------------------------------
 
-getNormalizedFilePath :: Monad m => Uri -> ExceptT String m NormalizedFilePath
-getNormalizedFilePath uri = handleMaybe errMsg
+getNormalizedFilePath :: Monad m => Uri -> ExceptT PluginError m NormalizedFilePath
+getNormalizedFilePath uri = handleMaybe (PluginUriToNormalizedFilePath uri)
         $ uriToNormalizedFilePath
         $ toNormalizedUri uri
-    where
-        errMsg = T.unpack $ "Failed converting " <> getUri uri <> " to NormalizedFilePath"
 
 -- ---------------------------------------------------------------------
-throwPluginError :: Monad m => String -> ExceptT String m b
-throwPluginError = throwE
+
+type PluginHandler e m a = ExceptT e m a
+
+pluginResponse :: Monad m => ExceptT PluginError m a -> m (Either ResponseError a)
+pluginResponse =
+  fmap (first handlePluginError)
+    . runExceptT
+
+pluginResponse' :: Monad m => (e -> ResponseError) -> ExceptT e m a -> m (Either ResponseError a)
+pluginResponse' handleError =
+  fmap (first handleError)
+    . runExceptT
+
+pluginResponseM :: Monad m => (t -> m (Either a b)) -> ExceptT t m b -> m (Either a b)
+pluginResponseM handler act =
+    runExceptT act >>= \case
+        Right r  -> pure $ Right r
+        Left err -> handler err
+
+handlePluginError :: PluginError -> ResponseError
+handlePluginError msg = ResponseError InternalError (prettyPluginError msg) Nothing
+
+data PluginError
+  = PluginInternalError
+  | PluginUriToFilePath J.Uri
+  | PluginUriToNormalizedFilePath J.Uri
+  | PluginErrorMessage T.Text
+
+prettyPluginError :: PluginError -> T.Text
+prettyPluginError = \case
+  PluginInternalError -> "Internal Plugin Error"
+  PluginUriToFilePath uri -> "Failed to translate URI " <> T.pack (show uri)
+  PluginUriToNormalizedFilePath uri -> "Failed converting " <> getUri uri <> " to NormalizedFilePath"
+  PluginErrorMessage msg -> "Plugin failed: " <> msg
+
+mkPluginErrorMessage :: T.Text -> PluginError
+mkPluginErrorMessage = PluginErrorMessage
+
+mkSimpleResponseError :: T.Text -> ResponseError
+mkSimpleResponseError err = ResponseError InternalError err Nothing
 
 handleMaybe :: Monad m => e -> Maybe b -> ExceptT e m b
 handleMaybe msg = maybe (throwE msg) return
 
 handleMaybeM :: Monad m => e -> m (Maybe b) -> ExceptT e m b
 handleMaybeM msg act = maybeM (throwE msg) return $ lift act
 
-pluginResponse :: Monad m => ExceptT String m a -> m (Either ResponseError a)
-pluginResponse =
-  fmap (first (\msg -> ResponseError InternalError (fromString msg) Nothing))
-    . runExceptT
+withError :: Functor m => (e' -> e) -> ExceptT e' m a -> ExceptT e m a
+withError = withExceptT
```

Comment:
Again, I think I'm broadly in favour of getting people to use "normal" `ExceptT` functions as much as possible.